rest-service/manager_rest/upload_manager.py

```diff
@@ -464,6 +474,165 @@ def _zip_dir(cls, dir_to_zip, target_zip_path):
+            new_blueprint.updated_at = now
+            new_blueprint.main_file_name = None
+            new_blueprint.visibility = visibility
+            new_blueprint.state = BlueprintUploadState.PENDING
+            sm.update(new_blueprint)
+        else:
+            new_blueprint = sm.put(Blueprint(
+                plan=None,
+                id=data_id,
+                description=None,
+                created_at=now,
+                updated_at=now,
+                main_file_name=None,
+                visibility=visibility,
+                state=BlueprintUploadState.PENDING
+            ))
+
+        if not blueprint_url:
+            new_blueprint.state = BlueprintUploadState.UPLOADING
+            sm.update(new_blueprint)
+            try:
+                self.upload_archive_to_file_server(data_id)
+            except Exception as e:
+                raise e
+
+        try:
+            get_resource_manager().upload_blueprint(
+                data_id,
+                args.application_file_name,
+                blueprint_url,
+                config.instance.file_server_root,   # for the import resolver
+            )
+        except manager_exceptions.ExistingRunningExecutionError as e:
+            new_blueprint.state = BlueprintUploadState.FAILED_UPLOADING
+            new_blueprint.error = str(e)
+            new_blueprint.error_traceback = traceback.format_exc()
+            sm.update(new_blueprint)
+            self.cleanup_blueprint_archive_from_file_server(
+                data_id, current_tenant.name)
+            raise e
+
+        return new_blueprint, 201
+
+    def upload_archive_to_file_server(self, blueprint_id):
+        file_server_root = config.instance.file_server_root
+        archive_target_path = tempfile.mktemp()
+        try:
+            self._save_file_locally_and_extract_inputs(
+                archive_target_path,
+                None,
+                self._get_kind())
+            dest_file_name = None
+            if not os.path.isfile(archive_target_path):
+                # If the archive is a folder, we're copying its content,
+                #  so there is no meaning to a specific archive file name...
+                # Relevant only in UploadedBlueprintsDeploymentUpdateManager
+                #  and UploadedPluginsManager
+                dest_file_name = None
+            self._move_archive_to_uploaded_dir(
+                blueprint_id,
+                file_server_root,
+                archive_target_path,
+                dest_file_name=dest_file_name)
+        except Exception as e:
+            sm = get_resource_manager().sm
+            blueprint = sm.get(Blueprint, blueprint_id)
+            blueprint.state = BlueprintUploadState.FAILED_UPLOADING
+            blueprint.error = str(e)
+            sm.update(blueprint)
+            self.cleanup_blueprint_archive_from_file_server(
+                blueprint_id, blueprint.tenant.name)
+            raise e
```

Comment:
note that `raise e` eats the traceback up to this point, so please avoid using `raise e` where the traceback could've been useful
instead, use just `raise` which re-raises the current exception with the traceback intact